_posts/2024-02-20-Navigating-Dependency-Chaos-with-Lockfiles.md

```diff
@@ -0,0 +1,348 @@
+
+## The dependency wilderness
+
+<!-- set a max width for mermaid diagram below so it doesn't render so large -->
+<style>
+.mermaid {
+  display: block;
+  margin: 0 auto;
+  max-height: 400px;
+}
+</style>
+
+```mermaid!
+gantt
+    your project :a, 2024-07-01, 3w
+    pandas :active ,b, 2024-07-07, 1w
+    numpy :active, c, 2024-07-09, 10d
+    scipy :active, d, 2024-07-04, 6d
+```
+
+_Dependencies are often on their own unpredictable schedule outside of your project's control._
+{:.center}
+
+
+Using existing software package dependencies helps conserve resources but comes with unique challenges related to unpredictability (such as when those dependencies are updated).
+This unpredictability can sometimes result in what's colloquially called ["dependency hell" or "dependency chaos"](https://en.wikipedia.org/wiki/Dependency_hell), where for example multiple external dependencies conflict with one another and are unable to be automatically resolved (among other issues).
+These challenges can be especially frustrating due to when they occur (often outside of our personal schedule awareness) and how long they can take to debug (finding fixes sometimes entails costly [trial-and-error](https://en.wikipedia.org/wiki/Trial_and_error)).
+It can feel like walking through a forest at night without a flashlight, constantly tripping over roots or running into stumps and branches!
+
+## Illuminating the dependency thicket
+
+```mermaid!
+flowchart LR
+
+ invention["Overwhelming\ninvention"]
+ dependencychaos["Infinite\ndependency\nchaos"]
+
+ invention <-......-> | software dependency \n choices |dependencychaos
+
+```
+
+_Software dependency choices may be understood through careful consideration between the cost of internal overwhelming invention vs external dependency chaos._
+{:.center}
+
+Dependency chaos can sometimes lead to ["not invented here syndrome"](https://en.wikipedia.org/wiki/Not_invented_here) where there's less trust in external-facing work outside of an individual or group of people.
+When or if this happens it can be important to understand dependencies as a __scale of choices between overwhelming invention and infinite dependency chaos__.
+For example, to accomplish a small project it may not be wise to create a brand new programming language (towards the extreme of overwhelming invention).
+On the other hand, if we depended upon all existing work within a certain context the solution may not be specialized, efficient, or resourceful enough to meet the goals within a reasonable amount of time.
+
+```mermaid!
+mindmap
+  root((Project))
+    Data storage
+      File 1
+      Database 2
+    Data x processing
+      Package X
+      Package Y
+    Integration
+      Solution A
+      Platform B
+
+```
+
+_Dependency awareness and opportunity can be grouped into concerns and documented as part of a literature review (seen here as a [mind map](https://en.wikipedia.org/wiki/Mind_map))._
+{:.center}
```

Comment:
Dang, you're right, it appears the same for me. Instinct here has me thinking this might have something to do with Mermaid versions, but not sure. I'll adjust this to get it displayed correctly.

Reply:
Awesome, looks like that worked; it's showing up fine in the Netlify preview for me now, at least. Kudos!

Reply:
This is fixed for the post but we may want to raise this with Jekyll Spaceship or elsewhere depending on what might be happening. I'm pretty sure it's a Mermaid version reference that makes this happen (mindmaps were I believe experimental or non-existent at one point). @vincerubinetti would you have any recommendations on this front?

Reply:
Perhaps it's worth raising an issue, though the jekyll-spaceship repo seems to have been inactive since 2022. Looking through the repo a bit, it looks like it just uses "https://mermaid.ink/svg/" as a conversion service (I think you brought this up before?). And [according to that service](https://github.com/jihchi/mermaid.ink/releases), it has already been using Mermaid 10 for a while. [Mermaid itself says that a mindmap is experimental](https://mermaid.js.org/syntax/mindmap.html) and doesn't seem to mention which version it came to, so maybe it's extremely recent and mermaid.ink is a little behind? Either way, it seems more like an issue with mermaid.ink or mermaid itself.

Also, appropriate that you'd run into a versioning issue on this blog post.

Reply:
Thanks @vincerubinetti ! And agreed on the versioning issue within the context of this post 😆 - fitting ! Maybe we need a lockfile update somewhere? 🙂 

I tried to create a mermaid.ink mindmap and it worked alright ([link here](https://mermaid.ink/svg/pako:eNpdkU1rwzAMhv-K0KmFUOJ8tc1toxvsMCj0so1ctETJvMV2cBxoV_rf5yRkjN38Pn4kJPuKpakYc1RSV4q6QgNYY9xqdbTmk0u3Xo8I4ECOoHfGUsMzAXiULYNY0mi8U88Q_ak4Q2dNyX0vdbOIRyq_fBd4-Q9eZ_CkHTeWnDR6MU6mHcYMd781LbnaWAX3GKBiq0hWfo3reF-g-2DFBeb-WHFNQ-sKLPTNq0NXkeOHSvpVMHd24ABpcOZ00eWSZ-cgyU-hMK-p7T3tSL8ZoxbJR8yveMZcbDciSdIwFlGY7PfRLsCLp2m4iTIRp_FOZJkIs-0twO-pgQiQpwGe57efvuD2A8EXeWs)). I'll give this a further investigation with a Jekyll-spaceship test when there's a moment.

Reply:
I've recreated the challenge outside of lab-website-template and just opened an issue on this topic here: https://github.com/jeffreytse/jekyll-spaceship/issues/96